program/src/processor/harvest_holder_rewards.rs

```diff
@@ -21,11 +36,231 @@ use crate::instruction::accounts::{Context, HarvestHolderRewardsAccounts};
+
+    require!(
+        ctx.accounts.vault_authority.key == &vault_signer,
+        StakeError::InvalidAuthority,
+        "vault authority",
+    );
+
+    // holder rewards (for the vault token account)
+    // - owner must be the rewards program
+    // - derivation must match
+
+    require!(
+        ctx.accounts.holder_rewards.owner == &paladin_rewards_program_client::ID,
+        ProgramError::InvalidAccountOwner,
+        "holder rewards",
+    );
+
+    let (derivation, _) = Pubkey::find_program_address(
+        &["holder".as_bytes(), ctx.accounts.vault.key.as_ref()],
+        &paladin_rewards_program_client::ID,
+    );
+
+    require!(
+        ctx.accounts.holder_rewards.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "holder rewards",
+    );
+
+    // Determine the holder rewards.
+
+    let holder_rewards = HolderRewards::try_from(ctx.accounts.holder_rewards)?;
+    let rewards = calculate_eligible_rewards(
+        holder_rewards.last_accumulated_rewards_per_token,
+        stake.last_seen_holder_rewards_per_token(),
+        stake.amount,
+    )?;
+
+    // Withdraw the holder rewards to the destination account.
+
+    if rewards != 0 {
+        // update the last seen holder rewards
+        stake.set_last_seen_holder_rewards_per_token(
+            holder_rewards.last_accumulated_rewards_per_token,
+        );
+
+        // Rewards are stored on the `vault` token account. We need to first withdraw the excess lamports
+        // from the vault token account to the vault authority account. Then transfer the rewards amount
+        // to the destination account and send the remaining excess lamports back to the vault token
+        // account.
+
+        let withdraw_ix = withdraw_excess_lamports(
+            ctx.accounts.token_program.key,
+            ctx.accounts.vault.key,
+            ctx.accounts.vault_authority.key,
+            ctx.accounts.vault_authority.key,
+            &[],
+        )?;
+
+        let signer_seeds = get_vault_pda_signer_seeds(ctx.accounts.config.key, &signer_bump);
+        // Stores the current lamports of the vault authority account before the withdraw so
+        // we can calculate the amount of rewards withdrawn.
+        let current_lamports = ctx.accounts.vault_authority.lamports();
+
+        invoke_signed(
+            &withdraw_ix,
+            &[
+                ctx.accounts.token_program.clone(),
+                ctx.accounts.vault.clone(),
+                ctx.accounts.vault_authority.clone(),
+            ],
+            &[&signer_seeds],
+        )?;
+
+        // If the withdraw does not result in enough lamports to cover the rewards, only
+        // harvest the available lamports. This should never happen, but the check is a
+        // failsafe.
+        let rewards = std::cmp::min(
+            rewards,
+            ctx.accounts
+                .vault_authority
+                .lamports()
+                .saturating_sub(current_lamports),
+        );
```

Comment:
It will invoke the `transfer` with `0`, which essentially will do nothing. In practice, this should not happen since we checked that we have rewards to distribute. We perhaps can add a log message?